### PR TITLE
Implementing Auto Hyperparameter Tuning Into Framework

### PR DESCRIPTION
**Pull request by aiden-dm**
This PR implements auto hyperparameter tuning logic for our framework. It takes a Random Search approach, selecting the parameter combination that results in the lowest value outputted by our custom loss function. I have added comments throughout my PR to ease the process of PR review!

### CODE REVIEW

File: Final_Structure/algo_args.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+def sample_args(method_name):
+    if method_name not in argspace_registry:
+        raise ValueError(f"ERROR: Method '{method_name}' not found in argspace_registry.")
+    space = argspace_registry[method_name]
+    return {k: random.choice(v) for k, v, in space.items()}
+
+def sample_args_ns(method_name):
+    params = sample_args(method_name)
+    return SimpleNamespace(**params)
```
Comment (aiden-dm) [inline Final_Structure/algo_args.py:41]: This file allows us to define the modifiable parameters for each unlearning algorithm. For each parameter we give a list of possible values from which the Random Search process can choose for a given iteration.

The sample_args() function randomly selects 1 value from each hyperparameter array for a specified unlearning method.

The sample_args_ns returns the randomly selected hyperparameter values as a SimpleNamespace, which is the hyperparameter input format expected by the unlearning algorithm implementations implemented thus far in the framework.

File: Final_Structure/badt.py [hunk 2]
```diff
@@ -52,6 +52,7 @@ def badt(full_model_path, loaders, args):
                         print_accuracies = args.print_accuracies)
 
     # Save a copy of the student model as a checkpoint
-    torch.save(student_model.state_dict(), args.check_path)
+    if args.save_checkpoint:
+        torch.save(student_model.state_dict(), args.check_path)
```
Comment (aiden-dm) [inline Final_Structure/badt.py:56]: Added the ability for the user to not save a checkpoint after the unlearning algorithm is done running. This is useful during the hyperparameter tuning process since many unnecessary checkpoints would be created otherwise.

File: Final_Structure/datasets.py [hunk 3]
```diff
@@ -111,34 +114,37 @@ def get_loaders(root, dataset, forget_classes, seed):
...
+    loaders = {
+        'train_loader': train_loader,
+        'valid_loader': valid_loader,
+        'test_loader': test_loader,
+        'train_forget_loader': train_forget_loader,
+        'train_retain_loader': train_retain_loader,
+        'valid_forget_loader': valid_forget_loader,
+        'valid_retain_loader': valid_retain_loader,
+        'test_forget_loader': test_forget_loader,
+        'test_retain_loader': test_retain_loader
```
Comment (aiden-dm) [inline Final_Structure/datasets.py:146]: Here I changed the output format of the get_loaders function. Beforehand it was a simple list, which meant that it had to be numerically indexed to extract individual loaders. A dictionary with string keys is much more readable and will lead to less confusion of our framework's users.

File: Final_Structure/training_params.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
+    return {
+        "model": model,
+        "optimizer": optimizer,
+        "criterion": nn.CrossEntropyLoss(),
+        "scheduler": torch.optim.lr_scheduler.ReduceLROnPlateau(
+            optimizer, mode='min', factor=0.1, patience=5, verbose=True
+        ),
+        "epochs": 65
+    }
```
Comment (aiden-dm) [inline Final_Structure/training_params.py:30]: This file defines the training parameters for a given model and dataset. This is used when you are training on the full train and on the forget train sets prior to running an experiment. By separating this setup into individual functions for different combinations of models and datasets, we make it significantly easier to expand in the future.

File: Final_Structure/tuning.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+            valid_forget_loader,
+            mia_seed=mia_seed
+        )
+
+        # Save best set of hyperparameters
+        if loss < best_loss:
+            best_loss = loss
+            best_config = args
+    
+    return best_config, best_loss
```
Comment (aiden-dm) [inline Final_Structure/tuning.py:102]: This is the file where I have implemented the Random Search hyperparameter tuning logic.

**Review by lilbhav — APPROVED** [review]
Makes sense, looks good!
